### PR TITLE
remove logical components, move hardware resources

### PR DESCRIPTION
**Pull request by Karsten1987**
Addresses the first block of #164 

The logical components such as Joint and Sensors are removed. There are only physical components now, namely Actuator, Sensor and System. These three are moved under the `components` namespace accordingly.

### CODE REVIEW

File: hardware_interface/include/hardware_interface/components/system.hpp
Comment (bmagyar) [inline hardware_interface/include/hardware_interface/components/system.hpp:41]: virtual destructor for a final class?

File: hardware_interface/src/components/sensor.cpp
Comment (bmagyar) [inline hardware_interface/src/components/sensor.cpp:26]: don't we only need this include where the macros are being used sir?

**Review by bmagyar — CHANGES_REQUESTED** [review]

**Review by bmagyar — APPROVED** [review]
Only two small comments, feel free to convert into follow-up issues if you prefer

**Comment by destogl** [comment]
Please wait for my review this evening... thanks!

File: hardware_interface/include/hardware_interface/components/system.hpp
Comment (Karsten1987) [inline hardware_interface/include/hardware_interface/components/system.hpp:41]:
```suggestion
  ~System() = default;
```

File: hardware_interface/src/components/actuator.cpp
Comment (Karsten1987) [inline hardware_interface/src/components/actuator.cpp:26]:
```suggestion
```

File: hardware_interface/src/components/sensor.cpp
Comment (Karsten1987) [inline hardware_interface/src/components/sensor.cpp:26]:
```suggestion
```

File: hardware_interface/src/components/system.cpp
Comment (Karsten1987) [inline hardware_interface/src/components/system.cpp:26]:
```suggestion
```

File: hardware_interface/src/components/sensor.cpp
Comment (Karsten1987) [reply hardware_interface/src/components/sensor.cpp]: I guess that's part of #199

Comment (bmagyar) [reply hardware_interface/src/components/sensor.cpp]: ok

**Review by bmagyar — APPROVED** [review]

**Review by destogl — COMMENTED** [review]
I have a few questions before merging:

1. I see there are read/write functions missing? How are you planning to enable those?
2. Since the hardware is parsing its interfaces, they should be somehow returned to the ResourceManager. I do not see those functions. Or are you planning to do something different?

**Comment by Karsten1987** [comment]
> I see there are read/write functions missing? How are you planning to enable those?

https://github.com/ros-controls/ros2_control/issues/164 depicts the roadmap I am planning to implement. The read functions are introduced in the second PR here: https://github.com/ros-controls/ros2_control/pull/203

> Since the hardware is parsing its interfaces, they should be somehow returned to the ResourceManager. I do not see those functions. Or are you planning to do something different?

The resource loaning is going to be implemented similar to #187 in which handles loaned to the controllers are unregistering themselves upon scope exit.

> We agreed, we want to enable some kind of logical components, how we should call them then? Just "Joint"/"Sensor"?

I think we agreed on the opposite to remove logical components (as the title of this PR says) and consider only hardware resources as components. The access to these handles is then done by re-using the existing "handles" we currently have.

File: hardware_interface/include/hardware_interface/components/sensor.hpp
Comment (Karsten1987) [reply hardware_interface/include/hardware_interface/components/sensor.hpp:61]: see #207

File: hardware_interface/include/hardware_interface/components/actuator_interface.hpp
Comment (Karsten1987) [reply hardware_interface/include/hardware_interface/components/actuator_interface.hpp:90]: see #203

Comment (destogl) [reply hardware_interface/include/hardware_interface/components/actuator_interface.hpp:90]: OK

File: hardware_interface/include/hardware_interface/components/sensor.hpp
Comment (destogl) [reply hardware_interface/include/hardware_interface/components/sensor.hpp:61]: OK

**Review by destogl — APPROVED** [review]
> I think we agreed on the opposite to remove logical components (as the title of this PR says) and consider only hardware resources as components. The access to these handles is then done by re-using the existing "handles" we currently have.

I think we are talking about the same thing, but currently I don't see anymore how we will implement those, or even better to say, how we should call them.

Just for the record:
- I am not 100% happy/convinced with renaming and removing "hardware" from the name, even though it is hardware representation on the end.
- I am afraid, we are getting to lock us in again as in ROS1

These are just my "feelings" and I cannot give you any sensible reason not to do this change, so I am approving this.

Thanks for the work and the explanations!

**Comment by destogl** [comment]
> I think we agreed on the opposite to remove logical components (as the title of this PR says) and consider only hardware resources as components. The access to these handles is then done by re-using the existing "handles" we currently have.

I think we are talking about the same thing, but currently I don't see anymore how we will implement those, or even better to say, how we should call them.

Just for the record:
- I am not 100% happy/convinced with renaming and removing "hardware" from the name, even though it is hardware representation on the end.
- I am afraid, we are getting to lock us in again as in ROS1

These are just my "feelings" and I cannot give you any sensible reason not to do this change, so I am approving this.
Thanks for the work and the explanations!

**Pull request merged**
